_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+  * Race conditions: if two clients send updates onto the same logical record, there is no guarantee that those updates will be applied ordered into the different datastores. That means that the app needs to implement itself Isolation from ACID properties, and that's not a trivial endeavour.
+  * Coupling: suddenly, what it was a simple app that writes to a database, it's a monster with coupling into different datastores. This approach doesn't scale organisationally as the owners of that service will have to serve requests for every single team.
+  * Crash recovery: even in the absence of concurrency this approach has an issue with Atomicity from ACID properties. What happens if the first updates succeed but the last one fails? It's not feasible to provide distributed heterogenous transactions, so, again, app will have to provide a do-it-yourself implementation of transactions.
+
+### Reflections on derived state
+
+Before of reaching to the proposed solution, let's think deeply about the concept of SSOT and derived data. Why do we keep different views of the original data? Let's see the reasoning behind different examples:
+
+  * Materialised views: our SSOT will have a format that won't satisfy every single query pattern. Maybe it's a normalised relational database as we need to save on storage costs and we were looking for strong referential constraints. We might have a multiple joins query that takes a lot of time, so we decided to persist that view on the database, to speed up reads. Database will take care of updating the materialised view as soon as the original tables gets updated.
+
+  * Indexes: every table, column family or whatever data store abstraction provides an efficient way of accessing its data. Maybe that's not enough for our query patterns, so we'll need to provide an index. Again, database will be in charge of updating it.
+
+  * Caches: some part of our data will be accessed more often. Also, there are physical storages that are faster, and more expensive on the other side. We use caches to keep that 'popular' data into efficient and spare data storage.
+
+  * Replicas: our state is the base of our business. We can't afford losing any data or being unavailable. The best way to protect ourselves from data corruption/loss or unavailable nodes is replicating our data to different locations.
+
+What is the state of our system? The original or the derived data. As long as the derived data has durability and consistency guarantees, they're a valid view of data at a given instant in time. What is the data structure used by databases to implement those derived forms? The answer is The Log. As we agreed that we want to use different data sources, specialised in their own non-functional domains, we could learn and use the lessons of The Log in order to provide reliable views of the state of our system.
+
+### The Log
```

Comment:
I like to give it a singular character as I'm talking about it like an abstraction. Like The Database, The Web Server, and not the databases.